### PR TITLE
Add Feature to develop

### PR DESCRIPTION
**Pull request by barakyoresh**

### CODE REVIEW

**Comment by dlahyani** [comment]
In general we avoid uploading 3rd party libraries to our projects repositories, instead we add them as submodules by referencing to their own repositories. Since you are using CocoaPods I think you are not even required to add them as submodules.

**Comment by dlahyani** [comment]
Round 1 complete.

File: ToneMapper/Model/Infastracture/TMGLBuffer.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+/// The size associated with this texture object.
+@property (readonly, nonatomic) CGSize size;
+
+/// Binds this framebuffer object in the opengl interface.
+- (void)bind;
+
+@end
+
+NS_ASSUME_NONNULL_END
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLBuffer.h]: New line.

File: ToneMapper/Model/Infastracture/TMGLBuffer.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+/// Protocol wrapping an opengl frame buffer object.
+@protocol TMGLBuffer <NSObject>
+
+/// The size associated with this texture object.
+@property (readonly, nonatomic) CGSize size;
+
+/// Binds this framebuffer object in the opengl interface.
+- (void)bind;
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLBuffer.h:12]: Methods before properties.

File: ToneMapper/Model/Infastracture/TMGLDrawer.m [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+- (instancetype)initWithProgramCompiler:(id<TMGLProgramCompiler>)programCompiler {
+  if (self = [super init]) {
+    _programCompiler = programCompiler;
+  }
+  return self;
+}
+
+- (void)drawTexture:(TMGLTexture *)texture toFrameBuffer:(id<TMGLBuffer>)frameBuffer
+withProgramParameters:(TMGLProgramParameters *)parameters {
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLDrawer.m]: 2 tabs

File: ToneMapper/Model/Infastracture/TMGLDrawer.m [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  glClearColor(kRed, kGreen, kBlue, kAlpha);
+  glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+  
+  if (!parameters) {
+    return;
+  }
+  
+  TMGLProgram *program =
+    [self.programCompiler createProgramFromProgramParams:parameters];
+  [program use];
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLDrawer.m:50]: :+1:

File: ToneMapper/Model/Infastracture/TMGLEngine.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/// Applies \c filter on the current workspace state texture, causing the current output texture
+/// to be the filtered version of the current workspace state.
+- (void)useFilter:(id<TMGLFilter>)filter;
+
+/// Draws the internal workspace state to the output buffer specified by \c outputBuffer method.
+- (void)draw;
+
+/// Sets the output buffer of the engine's subsequent draws as \c outputBuffer. \c outputBuffer's
+/// size property will also determine the aspect ratio correction made from image size to display.
+- (void)outputBuffer:(id <TMGLBuffer>)outputBuffer;
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLEngine.h]: Make this a property.

File: ToneMapper/Model/Infastracture/TMGLProgram/Internal Shaders/TMGLIdentityProgram.fsh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) 2015 Lightricks. All rights reserved.
+// Created by Barak Yoresh.
+
+varying highp vec2 textureCoordinates;
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLProgram/Internal Shaders/TMGLIdentityProgram.fsh]: Performance wise it is better to start with lowp variables and change them to highp only if you have any errors.

File: ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramParameters.m [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                                                               andVertexCount:_vertexCount];
+}
+
+#pragma mark -
+#pragma mark Equality
+#pragma mark -
+
+/// Cached equals created to check if this  \c TMGLProgramParameters is equal as far as the cache
+/// is concerned to \c other.
+- (BOOL)isEqual:(id)object {
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramParameters.m:76]: Such behavior should be documented.

File: ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramCacheManager.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+#import <OpenGLES/ES2/gl.h>
+
+@class TMGLProgramParameters, TMGLProgram;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// Program cache manager, saves handle for compiled programs and returns them if
+/// \c TMGLProgramParameters match a compiled handle and do not require recompiling.
+@interface TMGLProgramCacheManager : NSObject
```
Comment (dlahyani) [inline ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramCacheManager.h:14]: :+1:

File: ToneMapperTests/TMGLCachedProgramCompilerTest.m [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    GLuint program1 = [programCompiler createProgramFromProgramParams:params];
+    [builder addAttribute:@"attr3" valuePointer:(GLuint *)1 size:1 type:GL_FLOAT andStride:1];
+    params = [builder build];
+    GLuint program2 = [programCompiler createProgramFromProgramParams:params];
+    expect(@(program1)).toNot(equal(@(program2)));
+  });
+  
+});
+
+QuickSpecEnd
```
Comment (dlahyani) [inline ToneMapperTests/TMGLCachedProgramCompilerTest.m]: new line

File: ToneMapperTests/TMGLDefaultProgramCompilerTest.m [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                                     fragmentShaderName:@"invalid shader"
+                                     attributes:[[NSArray alloc] init]
+                                     unifroms:[[NSArray alloc] init] drawingMethod:0
+                                     andVertexCount:0];
+    expect([programCompiler createProgramFromProgramParams:params]).to(beNil());
+  });
+  
+});
+
+QuickSpecEnd
```
Comment (dlahyani) [inline ToneMapperTests/TMGLDefaultProgramCompilerTest.m]: same

File: ToneMapperTests/TMGLProgramParametersBuilderTest.m [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    it(@"Should be able to set a vertex count", ^{
+      expect(@(params.vertexCount)).to(equal(@1));
+    });
+    
+  });
+ 
+  
+});
+
+QuickSpecEnd
```
Comment (dlahyani) [inline ToneMapperTests/TMGLProgramParametersBuilderTest.m]: same

File: ToneMapperTests/TMGLProgramParametersTest.m [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+//      [MKTGiven([uniform name]) willReturn:@"uinf1"];
+//      TMGLProgramParameters *programParameters2 = [programParameters parametersWithReplacedUniform:unifrom2];
+//      expect([[programParameters2 uniforms] firstObject]).to(equal(unifrom2));
+    });
+    
+  });
+
+});
+
+QuickSpecEnd
```
Comment (dlahyani) [inline ToneMapperTests/TMGLProgramParametersTest.m]: same

**Comment by dlahyani** [comment]
Round 2. Fix these comments, fix the Uniform-Program interface and you can merge.

File: ToneMapper/Model/Infastracture/TMGLBuffer.h (hunk 1 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLBuffer.h]: done

File: ToneMapper/Model/Infastracture/TMGLBuffer.h (hunk 2 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLBuffer.h:12]: done

File: ToneMapper/Model/Infastracture/TMGLDrawer.m (hunk 3 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLDrawer.m]: done

File: ToneMapper/Model/Infastracture/TMGLDrawer.m (hunk 4 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLDrawer.m:50]: :smile:

File: ToneMapper/Model/Infastracture/TMGLEngine.h (hunk 5 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLEngine.h]: done

File: ToneMapper/Model/Infastracture/TMGLProgram/Internal Shaders/TMGLIdentityProgram.fsh (hunk 6 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLProgram/Internal Shaders/TMGLIdentityProgram.fsh]: done

File: ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramParameters.m (hunk 7 above)
Comment (barakyoresh) [inline ToneMapper/Model/Infastracture/TMGLProgram/TMGLProgramParameters.m:76]: done

**Pull request merged**